zypp/MediaSetAccess.cc

```diff
@@ -170,13 +171,23 @@ IMPL_PTR_TYPE(MediaSetAccess);
     return op.result;
   }
 
-  Pathname MediaSetAccess::provideFileFromUrl(const Url &file_url, ProvideFileOptions options)
+  ManagedFile MediaSetAccess::provideFileFromUrl(const Url &file_url, ProvideFileOptions options)
   {
     Url url(file_url);
     Pathname path(url.getPathName());
     url.setPathName ("/");
     MediaSetAccess access(url);
-    return access.provideFile(path, 1, options);
+
+    filesystem::TmpFile tmpFile;
+    tmpFile.autoCleanup(false); //cleaned up by ManagedFile
+
+    Pathname file = access.provideFile(path, 1, options);
+
+    //prevent the file from being deleted when MediaSetAccess gets out of scope
+    if ( filesystem::hardlinkCopy(file, tmpFile.path()) != 0 )
+      ZYPP_THROW(Exception("Can't copy file from " + file.asString() + " to " +  tmpFile.path().asString() ));
+
+    return ManagedFile ( tmpFile.path(), filesystem::unlink );
```

Comment:
IMO not exception safe. `tmpFile.autoCleanup(false)` must not be called before the `ManagedFile` actually manages it.

I'd almost consider providing a `static ManagedFile TmpFile::asManagedFile();` in `TmpPath.h/.cc` (as separate commit). Then there's no need for the code to fiddle with the `TmpFile` at all.

Reply:
Ouch yes, Done